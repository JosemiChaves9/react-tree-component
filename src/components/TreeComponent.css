.item {
  margin-left: 25px;
}

ul {
  list-style-type: none;
}
