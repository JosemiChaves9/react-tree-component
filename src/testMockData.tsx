export const data = [
  {
    name: 'Parent 1',
    children: [{ name: 'Children 1-1' }, { name: 'Children 1-2' }],
  },
  {
    name: 'Parent 2',
    children: [
      { name: 'Children 2-1' },
      { name: 'Children 2-2', children: [{ name: 'Children 2-2-1' }] },
    ],
  },
  {
    name: 'Parent 3',
    children: [
      {
        name: 'Children 3-1',
        rightContent: "I'm the children of Children 3-1",
      },
    ],
  },
];
