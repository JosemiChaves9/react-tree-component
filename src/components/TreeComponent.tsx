import './TreeComponent.css';
import {
  faChevronDown,
  faChevronRight,
} from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { useState } from 'react';

export interface DataItem {
  name: string;
  rightText?: string;
  children?: DataItem[];
  rightContent?: any;
}

export const TreeComponent = (props: { data: DataItem[] }) => {
  const [activeMenus, setActiveMenus] = useState<string[]>([]);
  const [toggleData, setToggleData] = useState<boolean>(false);

  const onClickOnArrow = (itemName: string) => {
    if (activeMenus) {
      if (activeMenus.includes(itemName)) {
        const index = activeMenus.indexOf(itemName);
        activeMenus.splice(index, 1);
      } else {
        setActiveMenus([...activeMenus, itemName]);
      }
    }
  };

  const renderChildren = (data: DataItem[]) => {
    return data.map((item: DataItem) => {
      return (
        <li
          unselectable={'off'}
          key={item.name}
          className='item'
          onClick={(e) => {
            onClickOnArrow(item.name);
            e.stopPropagation();
          }}>
          {item.children && activeMenus.includes(item.name) && (
            <FontAwesomeIcon
              icon={faChevronDown}
              className='downArrow'
              onClick={() => {
                setToggleData(!toggleData);
              }}
            />
          )}

          {item.children && !activeMenus.includes(item.name) && (
            <FontAwesomeIcon
              icon={faChevronRight}
              className='downArrow'
              onClick={() => {
                setToggleData(!toggleData);
              }}
            />
          )}
          {item.name}
          {item.rightContent && (
            <span className='rightContent'>{item.rightContent}</span>
          )}

          {activeMenus.includes(item.name) &&
            item.children &&
            renderChildren(item.children)}
        </li>
      );
    });
  };

  return (
    <>
      <ul>{renderChildren(props.data)}</ul>
    </>
  );
};
