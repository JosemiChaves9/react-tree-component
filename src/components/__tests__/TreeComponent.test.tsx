import { fireEvent, queryByText, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { data } from '../../testMockData';
import { TreeComponent } from '../TreeComponent';

it('First render correctly', () => {
  const tree = renderer.create(<TreeComponent data={data} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('In first render only the three first levels should appear.', () => {
  const { queryByText } = render(<TreeComponent data={data} />);

  expect(queryByText('Parent 1')).toBeDefined();
  expect(queryByText('Children 2-2')).toBeNull();
  expect(queryByText('Parent 3')).toBeDefined();
});

test('When clicked in Parent 1, should return Children 1-1 and Children 2-2, and if clicked again should remove it.', () => {
  const { getByText, queryByText } = render(<TreeComponent data={data} />);

  getByText('Parent 1').click();

  expect(getByText('Children 1-1')).toBeDefined();
  expect(getByText('Children 1-2')).toBeDefined();

  getByText('Parent 1').click();
  expect(queryByText('Children 1-1')).toBeNull();
  expect(queryByText('Children 1-2')).toBeNull();
});

test('When clicked the Parent 2, should return Children 2-1 and Children 2-2, finally clicking Children 2-2 will return Children 2-2-1.', () => {
  const { getByText, queryByText } = render(<TreeComponent data={data} />);

  getByText('Parent 2').click();
  expect(getByText('Children 2-1')).toBeDefined();
  expect(getByText('Children 2-2')).toBeDefined();
  expect(queryByText('Children 2-2-1')).toBeNull();

  getByText('Children 2-2').click();
  expect(getByText('Children 2-2-1')).toBeDefined();
});

test('In Children 3-1 should display a span that is passed as rightContent', () => {
  const { getByText, queryByText } = render(<TreeComponent data={data} />);

  expect(getByText('Parent 3')).toBeDefined();
  expect(queryByText('Children 3-1')).toBeNull();

  getByText('Parent 3').click();
  expect(getByText('Children 3-1')).toBeDefined();
  expect(getByText("I'm the children of Children 3-1")).toBeDefined();
});
