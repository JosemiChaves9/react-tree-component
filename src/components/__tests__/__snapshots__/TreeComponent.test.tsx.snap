// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`First render correctly 1`] = `
<ul>
  <li
    className="item"
    onClick={[Function]}
    unselectable="off"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-chevron-right fa-w-10 downArrow"
      data-icon="chevron-right"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 320 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    Parent 1
  </li>
  <li
    className="item"
    onClick={[Function]}
    unselectable="off"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-chevron-right fa-w-10 downArrow"
      data-icon="chevron-right"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 320 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    Parent 2
  </li>
  <li
    className="item"
    onClick={[Function]}
    unselectable="off"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-chevron-right fa-w-10 downArrow"
      data-icon="chevron-right"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 320 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    Parent 3
  </li>
</ul>
`;
